### PR TITLE
refactor the metrics

### PR DESCRIPTION
**Pull request by faneshion**
#312

### CODE REVIEW

File: matchzoo/metrics.py [hunk 1]
```diff
@@ -1,190 +1,210 @@
...
+            break
+        if label > rel_threshold:
+            s += (math.pow(2., label) - 1.) / math.log(2. + i)
+    return s
+
+
+def ndcg_at_k(y_true: list,
+              y_pred: list,
+              k: int,
+              rel_threshold: int=0) -> float:
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: `rel_threshold` -> `threshold`

File: matchzoo/metrics.py [hunk 2]
```diff
@@ -1,190 +1,210 @@
...
-            return r[0] + np.sum(r[1:] / np.log2(np.arange(2, r.size + 1)))
-        else:
-            return np.sum(r / np.log2(np.arange(2, r.size + 2)))
-    return 0.
-
-
-def ndcg_at_k(r: list, k: int, method: int=0) -> float:
+    if k <= 0.:
+        return 0.
+    c = sort_couple(y_true, y_pred)
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: what is `c`? use descriptive variable name

File: matchzoo/metrics.py [hunk 3]
```diff
@@ -1,190 +1,210 @@
...
+    s = 0.
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: what is `s`?

File: matchzoo/metrics.py [hunk 4]
```diff
@@ -1,190 +1,210 @@
...
+    for i, (label, score) in enumerate(c):
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: `i` -> `idx` or `index`

File: matchzoo/metrics.py [hunk 5]
```diff
@@ -1,190 +1,210 @@
...
+        if i >= k:
+            break
+        if label > rel_threshold:
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: threshold

File: matchzoo/metrics.py [hunk 6]
```diff
@@ -1,190 +1,210 @@
...
+        return 0.
+    else:
+        return s/ipos
 
 
-def dcg_at_k(r: list, k: int, method: int=0) -> float:
+def dcg_at_k(y_true: list,
+             y_pred: list,
+             k: int,
+             rel_threshold: int=0) -> float:
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: same

File: matchzoo/metrics.py [hunk 7]
```diff
@@ -1,190 +1,210 @@
...
+    :param y_true: The ground true label of each document.
+    :param y_pred: The predicted scores of each document.
+    :param rel_threshold: the label threshold of relevance degree.
     :return: Mean average precision.
     """
-    return np.mean([average_precision(r) for r in rs])
+    s = 0.
+    ipos = 0
+    c = sort_couple(y_true, y_pred)
+    for j, (label, score) in enumerate(c):
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: j

File: matchzoo/metrics.py [hunk 8]
```diff
@@ -1,190 +1,210 @@
...
 
-    :param rs: Iterator of relevance scores (list or numpy) in rank order
-               (first element is the first item).
+    :param y_true: The ground true label of each document.
+    :param y_pred: The predicted scores of each document.
+    :param rel_threshold: the label threshold of relevance degree.
     :return: Mean average precision.
     """
-    return np.mean([average_precision(r) for r in rs])
+    s = 0.
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: s

File: matchzoo/metrics.py [hunk 9]
```diff
@@ -1,190 +1,210 @@
...
+    ipos = 0
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: ipos?

File: matchzoo/metrics.py [hunk 10]
```diff
@@ -1,190 +1,210 @@
...
 import numpy as np
+import typing
 
 
-def mean_reciprocal_rank(rs: list) -> float:
-    """
-    Calculate reciprocal of the rank of the first relevant item.
+def _to_list(x: typing.Any) -> list:
+    """Transfer the inputs to list."""
+    if isinstance(x, list):
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: this function is useless if you clearly define x to be list or array-like objects in docstring

File: matchzoo/metrics.py [hunk 11]
```diff
@@ -1,190 +1,210 @@
...
+        return x
+    return [x]
 
-    First element is 'rank 1'. Relevance is binary (nonzero is relevant).
 
-    Example:
-        >>> rs = [[0, 0, 0, 1], [1, 0, 0], [1, 0, 0]]
-        >>> mean_reciprocal_rank(rs)
-        0.75
+def sort_couple(labels: list, scores: list) -> list:
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: sort and couple?

File: matchzoo/engine/base_model.py [hunk 12]
```diff
@@ -189,6 +189,43 @@ def evaluate(
...
+            :class:`engine.BaseGenerator`.
+        :param metrics: metrics used on evaluation.
+        :param batch_size: number of samples per gradient update
+        :return: list of scalars (losses as well as metrics). The
+            attribute `model.backend.metrics_names` will give you
+            the display labels for the scalar outputs.
+
+        """
+        qinfo = {}
+        for (x, y) in generator:
```
Comment (bwanglzu) [inline matchzoo/engine/base_model.py]: for batch in generator? I think `generator[0]` is the first batch, then you can access x and y through batch.

File: matchzoo/engine/base_model.py [hunk 13]
```diff
@@ -189,6 +189,43 @@ def evaluate(
...
+        :param generator: A generator, an instance of
+            :class:`engine.BaseGenerator`.
+        :param metrics: metrics used on evaluation.
+        :param batch_size: number of samples per gradient update
+        :return: list of scalars (losses as well as metrics). The
+            attribute `model.backend.metrics_names` will give you
+            the display labels for the scalar outputs.
+
+        """
+        qinfo = {}
```
Comment (bwanglzu) [inline matchzoo/engine/base_model.py]: not quite understand the meaning for `qinfo`

File: matchzoo/engine/base_model.py [hunk 14]
```diff
@@ -189,6 +189,43 @@ def evaluate(
...
+            y_pred = self._backend.predict(x=batch_x, batch_size=batch_size)
+            scores = zip(y, y_pred)
+            for idx, qid in enumerate(x['id_left']):
+                if qid not in query_lists:
+                    query_lists[qid] = []
+                query_lists[qid].append(scores[idx])
+        outs = {}
+        for metric in metrics:
+            outs[metric] = 0.
+            for qid, scores in query_lists.items():
```
Comment (bwanglzu) [inline matchzoo/engine/base_model.py]: `query_lists` is a dict?

Comment (faneshion) [reply matchzoo/engine/base_model.py]: This part will be refactored by Uduse, and you can ignore this PR.

**Pull request closed without merging**